File: 131-palindrome-partitioning/palindrome-partitioning.java
class Solution {
    List<List<String>> res = new ArrayList<>();
    List<String> l = new ArrayList<>();
    public List<List<String>> partition(String s) {
        dfs(s,0);
        return res;        
    }
    public boolean isSame(String q){
        int i=0,j = q.length()-1;
        while(i<j){
            if(q.charAt(i)!=q.charAt(j)) return false;
            i+=1;
            j-=1;
        }
        return true;


    }
    public void dfs(String s,int i){
        if(i==s.length()){
            res.add(new ArrayList(l));
            return ;
        }

        String q = "";
        for(int k = i;k<s.length();k++){
            q = q+s.charAt(k);
            if(isSame(q)){
                l.add(q);
                dfs(s,k+1);
                l.remove(l.size()-1);
            }
        }
        return ;
    }
}

File: 131-palindrome-partitioning/palindrome-partitioning.cpp
class Solution {
public:
    vector<vector<string>> res;
    vector<string> l;
    
    bool isSame(string q){
        int i=0,j=q.size()-1;
        while(i<j){
            if(q[i]!=q[j]) return false;
            i+=1;
            j-=1; 
        }
        return true;
    }
    void dfs(string s,int i){
        if(i==s.size()){
            res.push_back(vector<string>(l));
            return ;
        }

        string q ="";
        for(int k=i;k<s.size();k++){
            q = q+s[k];
            if(isSame(q)){
                l.push_back(q);
                dfs(s,k+1);
                l.pop_back();
            }
        }
        return;
    }
    vector<vector<string>> partition(string s) {
        dfs(s,0);
        return res;
    }
};